docs/spec.rst

```diff
@@ -0,0 +1,506 @@
+declarations involving unspecified nullness. Note that
+`supermethods <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=id.5nvbughni6vx>`__
+and
+`superparameters <https://docs.google.com/document/d/1KQrBxwaVIPIac_6SCf--w-vZBeHkTvtaqPSU_icIccc/edit#bookmark=id.m2gxs1ddzqwp>`__
+may be defined by members of parameterized supertypes.
+
+Intended “meaning”
+~~~~~~~~~~~~~~~~~~
+
+The broad intention of nullability is that type usages specified as
+non-null should at runtime only represent non-null values. Since fields
+have to be initialized before they can be non-null, this turns out to be
+impractical in edge cases, however. Further, this problem can affect
+method return values, e.g., when methods return field values. We
+therefore likewise only expect non-null guarantees to hold for instance
+(static, respectively) fields and method results by the time their
+declaring class’s constructor (static initializer, respectively) has
+finished (similar to when final fields are guaranteed to be
+initialized).
+
+Note this semantics does allow for situations in which null values may
+be observable in “non-null” fields and method returns, namely while
+objects are under construction. We simply encourage API owners to
+minimize these cases for non-private (both static and instance) fields
+and methods, which typically involves not “leaking” object references
+outside an API until they’re fully constructed. Some tools may attempt
+to identify such “leaks” and may attempt to ensure proper field
+initialization as defined here during object construction.
+
+Examples
+~~~~~~~~
+
+As an example, let’s consider a fragment of Guava’s ``ImmutableMap``:
+
+.. code:: java
+
+   @DefaultNonNull
+   public class ImmutableMap<K, V> implements Map<K, V> {
+     public static <K, V> ImmutableMap<K, V> of(K key, V value);
+     public @Nullable V get(@Nullable Object key);
+   }
+
+Because of the use of ``@DefaultNonNull``, every type use in this
+class’s declaration is fixed to either nullable or non-null (including
+type-variable uses, since their type parameters are considered
+implicitly bounded by non-null ``Object``).
+
+-  Can the parameters to ``of()``\ ’s be null? no (from ``K`` and
+   ``V``\ ’s bounds, which are determined implicitly by
+   ``@DefaultNonNull``)
+-  Can ``get()``\ ’s return ``null``? yes (from its explicit annotation)
+-  It is a mismatch to refer to
+   ``ImmutableMap<@Nullable String, Object>`` because
+   ``@Nullable String`` is outside of ``K``\ ’s bounds.
+
+To illustrate wildcards, consider a method return type
+``ImmutableMap<? extends @Nullable String, ?>`` with no defaulting
+annotation in scope:
+
+-  Can the method return a null map? That is unspecified (since no
+   defaulting annotation is in scope)
+-  Can the map’s keys be null? no. This is because the wildcard’s
+   explicit bound, ``@Nullable String``, is intersected with the
```

Comment:
Maybe this intersection should be mentioned in the `Type arguments of parameterized types` section? It's a bit surprising to first read about this in the example.

Reply:
Oh, yeah, that's in one of the parts I cut out. How about I cut this section down to refer only to the type parameters' bounds (see new version)? Or maybe we'd be better off without the section entirely?